packages/openai_ai_model/src/microsoft/teams/openai_ai_model/ai_model.py

```diff
@@ -0,0 +1,298 @@
+
+        # Push response to memory (only if not recursing)
+        await memory.push(model_response)
+        self.logger.debug("Chat Completions conversation completed, returning response")
+
+        return model_response
+
+    async def _execute_functions(
+        self, input: Message, functions: dict[str, Function[BaseModel]] | None
+    ) -> list[FunctionMessage]:
+        """Execute any pending function calls in the input message."""
+        function_results: list[FunctionMessage] = []
+
+        if isinstance(input, ModelMessage) and input.function_calls:
+            # Execute any pending function calls
+            self.logger.debug(f"Executing {len(input.function_calls)} function calls")
+            for call in input.function_calls:
+                if functions and call.name in functions:
+                    function = functions[call.name]
+                    try:
+                        # Parse arguments into Pydantic model
+                        parsed_args = function.parameter_schema(**call.arguments)
+
+                        # Handle both sync and async function handlers
+                        result = function.handler(parsed_args)
+                        if inspect.isawaitable(result):
+                            fn_res = await result
+                        else:
+                            fn_res = result
+
+                        # Create function result message
+                        function_results.append(FunctionMessage(content=fn_res, function_id=call.id))
+                    except Exception as e:
+                        # Handle function execution errors
+                        function_results.append(
+                            FunctionMessage(content=f"Function execution failed: {str(e)}", function_id=call.id)
+                        )
+
+        return function_results
+
+    async def _handle_streaming_response(
+        self, stream: AsyncStream[ChatCompletionChunk], on_chunk: Callable[[str], Awaitable[None]] | None
+    ) -> ModelMessage:
+        """Handle streaming OpenAI response and accumulate into ModelMessage."""
+        # Initialize accumulation structures
+        content = ""
+        tool_calls_data: list[_ToolCallData] = []  # List of dict to accumulate tool call data
+
+        async for chunk in stream:
+            if not chunk.choices:
+                continue
+
+            delta = chunk.choices[0].delta
+
+            # Handle content streaming
+            if delta.content:
+                content += delta.content
+                if on_chunk:
+                    await on_chunk(delta.content)
+
+            # Handle tool calls streaming
+            if delta.tool_calls:
+                for call_delta in delta.tool_calls:
+                    # Ensure we have enough slots for this index
+                    while len(tool_calls_data) <= call_delta.index:
+                        tool_calls_data.append({"id": "", "name": "", "arguments_str": ""})
```

Comment:
nit: can we use `ToolCallData` dict defined above

Reply:
We are aren't we?